<?xml version="1.0"?>

<launch>

    <!-- ARGS -->
    <arg name="rvizfile"    default = "$(find gtsam_backend)/../../rviz/mvsec_test.rviz" />
    <arg name="config_path" default = "$(find gtsam_backend)/../../config/mvsec_vins_config.yaml" />
    <arg name="vins_folder" default = "$(find feature_tracker)/../" />

    <!-- RVIZ -->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(arg rvizfile)"/>

    <!-- FRONT END! -->
    <node name="feature_tracker" pkg="feature_tracker" type="feature_tracker" output="log">
      <param name="config_file" type="string" value="$(arg config_path)" />
      <param name="vins_folder" type="string" value="$(arg vins_folder)" />
    </node>

    <!-- MASTER NODE! -->
    <node pkg="gtsam_backend" type="vio" name="vio" output="screen" clear_params="true" required="true">
      <remap from="/vio/data_imu" to="/sensors/applanix/imu_data" />
      <remap from="/vio/data_uv"  to="/feature_tracker/feature" />

      <param name="fixedId"  type="string" value="world" />
      <rosparam param="gravity">[0.0, 0.0, 0.0]</rosparam>
      <param name="imuWait"  type="int" value="1096" />
      <param name="featWait" type="int" value="5" />
      
      <!-- Rotation and Translation from camera frame to IMU frame -->
      <rosparam param="R_C0toI">[ -0.0389980, -0.9992386, -0.0012147,
                                   0.0110205,  0.0007855, -0.9999390,
                                   0.9991785, -0.0390090,  0.0109815 ]</rosparam>
      <rosparam param="p_IinC0">[ 0.98, 0.0, 0.52]</rosparam>

      <!-- Initialization -->
      <rosparam param="prior_qGtoI">[0.0, 0.0, 0.0, 1.0]</rosparam>       
      <rosparam param="prior_pIinG">[0.0, 0.0, 0.0]</rosparam>       
      <rosparam param="prior_vIinG">[0.0, 0.0, 0.0]</rosparam>       
      <rosparam param="prior_ba">[0.00489771688759235973,  0.00800897351104824969, 0.03020588299505782420]</rosparam>       
      <rosparam param="prior_bg">[-0.00041196751646696610, 0.00992948457018005999, 0.02188282212555122189]</rosparam>       

      <param name="sigma_camera"                 type="double"   value="0.206555403" />
      
      <!-- Noise Values for ETH Dataset Sensor -->
      <param name="accelerometer_noise_density"  type="double"   value="0.08" />    <!-- sigma_a -->
      <param name="gyroscope_noise_density"      type="double"   value="0.004" />   <!-- sigma_g -->
      <param name="accelerometer_random_walk"    type="double"   value="0.0004" /> <!-- sigma_wa -->
      <param name="gyroscope_random_walk"        type="double"   value="2.0e-5" />  <!-- sigma_wg -->

      <!-- Noise Values for Initialization -->
      <param name="sigma_prior_rotation"         type="double"   value="0.001" />
      <param name="sigma_prior_translation"      type="double"   value="0.001" />
      <param name="sigma_velocity"               type="double"   value="0.1" />
      <param name="sigma_bias"                   type="double"   value="0.15" />
      <param name="sigma_pose_rotation"          type="double"   value="0.1" />
      <param name="sigma_pose_translation"       type="double"   value="0.2" />
    </node>

    <node name="camera_info_relay" pkg="topic_tools" type="relay" args="/sensors/broadrreachcam_front/camera_info /feature_tracker/camera_info" />
    <node pkg="rosbag" type="play" name="player" output="screen"
          args="-q $(find feature_tracker)/../../bags/20190614-101546_processed_cut.bag /tf:=/tf_dev_null" />
    <node name="republish_left_image" type="republish" pkg="image_transport" output="screen"
          args="compressed in:=/sensors/broadrreachcam_front/image_compressed raw out:=/sensors/broadrreachcam_front/image" />
</launch>
